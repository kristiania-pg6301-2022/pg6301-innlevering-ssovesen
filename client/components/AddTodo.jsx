import { useState } from "react";
import React from "react";
import { requestJSON } from "../utils/httpRequestHandlers";

export function AddTodo({ getTodo, reload }) {
  const [title, setTitle] = useState("");
  const [text, setText] = useState("");

  async function postTodo(e) {
    e.preventDefault();
    const id = Math.floor(Math.random() * 10000);
    await requestJSON("/api/addTodo", { id, title, text }, "post");
    reloadTodo();
  }

  const reloadTodo = () => {
    setTitle("");
    setText("");
    reload();
  };

  return (
    <form className="todoForm" onSubmit={postTodo}>
      <div>
        Title:{" "}
        <input
          type="text"
          name="title"
          value={title}
          onChange={(e) => setTitle(e.target.value)}
        />
      </div>
      <div>
        Text{" "}
        <input
          type="text"
          name="text"
          value={text}
          onChange={(e) => setText(e.target.value)}
        />
      </div>
      <div>
        <button type="submit">Add</button>
      </div>
    </form>
  );
}
