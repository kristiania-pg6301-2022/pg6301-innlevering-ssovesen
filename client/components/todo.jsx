import React from "react";
export const Todo = ({ todo }) => {
  const { title, text, completed } = todo;

  return (
    <li>
      <h1>{title}</h1>
      <p>{text}</p>
    </li>
  );
};
