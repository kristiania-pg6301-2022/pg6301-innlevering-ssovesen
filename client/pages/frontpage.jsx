import React from "react";

export function Frontpage() {
  return <h1>FrontPage</h1>;
}
