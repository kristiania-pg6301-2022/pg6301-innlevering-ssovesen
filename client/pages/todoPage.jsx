import React, { useEffect, useState } from "react";
import { Todo } from "../components/todo";
import { AddTodo } from "../components/AddTodo";
import { useLoader } from "../utils/hooks/useLoader";
import {
  fetchJSON,
  deleteJSON,
  requestJSON,
} from "../utils/httpRequestHandlers";

export function TodoPage() {
  const {
    loading,
    error,
    data: todoList,
    reload,
  } = useLoader(async () => await fetchJSON("/api/getAll"));

  async function onDelete(todo) {
    await deleteJSON(`/api/todo/${todo.id}`);
    await reload();
  }

  async function onComplete(todo) {
    let status;
    todo.completed ? (status = false) : (status = true);
    await requestJSON(`/api/todo/${todo.id}`, { completed: status }, "PUT");
    await reload();
  }

  return (
    <>
      <h1>Welcome to this awesome todo app</h1>
      {loading && <p>Loading...</p>}
      {error && <p>Error: {error.message}</p>}
      <div className="app">
        <AddTodo getTodo={todoList} reload={reload} />
        <div>
          {
            <ul>
              {todoList?.map((todo) => {
                if (!todo.completed) {
                  return (
                    <Todo
                      key={todo.id}
                      todo={todo}
                      whenDeleted={onDelete}
                      whenCompleted={onComplete}
                    />
                  );
                }
              })}
            </ul>
          }
          <h1>Completed</h1>
          <ul>
            {todoList?.map((todo) => {
              if (todo.completed) {
                return (
                  <Todo
                    key={todo.id}
                    todo={todo}
                    whenDeleted={onDelete}
                    whenCompleted={onComplete}
                  />
                );
              }
            })}
          </ul>
        </div>
      </div>
    </>
  );
}
