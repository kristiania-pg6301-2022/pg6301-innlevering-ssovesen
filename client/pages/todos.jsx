import React, { useEffect, useState } from 'react';
import { Todo } from '../components/todo';
import { AddTodo } from '../components/AddTodo';

export function Todos() {
	const [todos, setTodos] = useState([]);

	function onDelete(todo) {
		console.log(todo, 'todo');
		fetch(`/api/todo/${todo.id}`, { method: 'DELETE' }).then(response =>
			response.json().then(data => console.log(data))
		);
		getTodos();
	}

	function onComplete(todo) {
		let status = undefined;
		todo.completed ? (status = false) : (status = true);

		console.log('onComplete works', todo.completed);

		const requestOptions = {
			method: 'Put',
			headers: { 'Content-Type': 'application/json' },
			body: JSON.stringify({ completed: status }),
		};

		fetch(`/api/todo/${todo.id}`, requestOptions).then(response =>
			response.json().then(data => console.log(data))
		);
		getTodos();
	}

	const getTodos = () => {
		fetch('/api/getAll')
			.then(response => response.json())
			.then(data => setTodos(data));
	};

	useEffect(() => {
		getTodos();
	}, []);

	return (
		<>
			<AddTodo getTodo={getTodos} />
			<div>
      {
				<ul>
					{todos.length > 0
						? todos.map(todo => {
								if (!todo.completed) {
									return (
										<Todo
											key={todo.id}
											todo={todo}
											whenDeleted={onDelete}
											whenCompleted={onComplete}
										/>
									);
								}
						  })
						: 'Du har ingen gjøremål'}
				</ul>
			}
			<h1>Completed</h1>
			<ul>
				{todos.length > 0
					? todos.map(todo => {
							if (todo.completed) {
								return (
									<Todo
										key={todo.id}
										todo={todo}
										whenDeleted={onDelete}
										whenCompleted={onComplete}
									/>
								);
							}
					  })
					: ''}
			</ul>
      </div>
		</>
	);
}
