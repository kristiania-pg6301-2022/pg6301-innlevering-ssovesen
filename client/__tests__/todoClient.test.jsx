import React from "react";
import ReactDOM from "react-dom";
import pretty from "pretty";
import { MemoryRouter } from "react-router-dom";
import { Frontpage } from "../pages/frontpage";
import { Todo } from "../components/todo";
import { TodoPage } from "../pages/todoPage";
import { Simulate } from "react-dom/test-utils";
import { AddTodo } from "../components/AddTodo";

describe("Client tests", () => {
  it("should display page ", () => {
    const element = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <Frontpage />
      </MemoryRouter>,
      element
    );
    expect(pretty(element.innerHTML)).toMatchSnapshot();
    ReactDOM.unmountComponentAtNode(element);
  });

  it("should render todos page", () => {
    const element = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter initialEntries={["/todos"]}>
        <TodoPage />
      </MemoryRouter>,
      element
    );
    expect(pretty(element.innerHTML)).toMatchSnapshot();
  });

  it("should complete todo", () => {
    const todo = {
      title: "Test Todo",
      text: "Test Description",
      completed: false,
    };
    const onComplete = jest.fn();
    const onDelete = jest.fn();

    const element = document.createElement("div");

    ReactDOM.render(<Todo todo={todo} whenCompleted={onComplete} />, element);
    console.log(element.querySelector("button"));
    Simulate.click(element.querySelector("button"));
    expect(onComplete).toBeCalled();
  });

  it("should delete todo", () => {
    const todo = {
      title: "Test Todo",
      text: "Test Description",
      completed: false,
    };
    const onDelete = jest.fn();

    const element = document.createElement("div");

    ReactDOM.render(<Todo todo={todo} whenDeleted={onDelete} />, element);
    Simulate.click(element.querySelector("button").nextSibling);
    expect(onDelete).toBeCalled();
  });

  /*  it("should add new todo", () => {
    const newTodo = { title: "New Todo", text: "New Text" };
    const postTodo = jest.fn();
    const element = document.createElement("div");
    ReactDOM.render(<AddTodo />, element);
    Simulate.submit(element.querySelector("form"), { newTodo });
    expect(postTodo).toBeCalled();
  }); */
});
