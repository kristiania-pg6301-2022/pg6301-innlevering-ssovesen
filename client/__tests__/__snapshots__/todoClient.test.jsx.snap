// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client tests should display page  1`] = `"<h1>FrontPage</h1><a href=\\"/todos\\">Todo</a>"`;

exports[`Client tests should render todos page 1`] = `
"<h1>Welcome to this awsome todo app</h1>
<p>Loading...</p>
<form class=\\"todoForm\\">
  <div>Title: <input type=\\"text\\" name=\\"title\\" value=\\"\\"></div>
  <div>Text <input type=\\"text\\" name=\\"text\\" value=\\"\\"></div>
  <div><button type=\\"submit\\">Add</button></div>
</form>
<ul></ul>
<h1>Completed</h1>
<ul></ul>"
`;
